File: prep/implementation/practice_iterations/two_pointers_5/TripletSumCloseToTarget.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

// O(n2) / O(1)
public class TripletSumCloseToTarget {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        int s = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        Arrays.sort(a);

        int minDiff = Integer.MAX_VALUE;
        for (int i = 0; i < n - 2; i++) {
          int left = i + 1, right = n - 1;

          while (left < right) {
            int diff = s - a[i] - a[left] - a[right];

            if (Math.abs(diff) < Math.abs(minDiff) ||
                (Math.abs(diff) == Math.abs(minDiff) && diff > minDiff)) {
              minDiff = diff;
            }

            if (diff > 0) {
              left++;
            } else {
              right--;
            }
          }
        }

        System.out.println((s - minDiff));
      }
    }
  }
}

File: lc/coreprep/implementation/practice_iterations/iteration_archives/two_pointers_3/triplet_sum_close_to_target.cpp
#include <algorithm>
#include <iostream>
#include <limits>
#include <vector>

using namespace std;

// O(n2) / O(n)
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, s;
  cin >> tt;

  while (tt--) {
    cin >> n >> s;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    sort(a.begin(), a.end());

    int min_diff = numeric_limits<int>::max();
    for (int i = 0; i < n - 2; i++) {
      int left = i + 1, right = n - 1;

      while (left < right) {
        int diff = s - a[i] - a[left] - a[right];

        if (abs(diff) < abs(min_diff) ||
            (abs(diff) == abs(min_diff) && diff > min_diff)) {
          min_diff = diff;
        }

        if (diff > 0) {
          left++;
        } else {
          right--;
        }
      }
    }

    cout << (s - min_diff) << "\n";
  }

  return 0;
}